src/main/java/com/redhat/devtools/intellij/knative/utils/WatchHandler.java

```diff
@@ -46,32 +49,47 @@ public static WatchHandler get(Kn kn) {
         return instance;
     }
 
-    public void watchResource(String id, String kindToWatch, Runnable doExecute) {
-        if (!watches.containsKey(id)) {
-            Watch watch = watchResource(kindToWatch, doExecute);
-            if (watch != null) {
-                watches.put(id, watch);
+    public void watchResource(String id, String kindToWatch, Runnable doExecute) throws IOException {
+        if (!watches.containsKey(id)
+            && (!watchRetry.containsKey(id) || watchRetry.get(id) < 3)) {
+            try {
+                Watch watch = watchResource(kindToWatch, doExecute);
+                if (watch != null) {
+                    watches.put(id, watch);
+                }
+            } catch (IOException e) {
+                int retry = watchRetry.getOrDefault(id, 0);
+                watchRetry.put(id, ++retry);
+                if (retry > 2) {
+                    throw new IOException("Unable to watch " + kindToWatch + " resources on cluster");
+                }
             }
         }
     }
 
-    private Watch watchResource(String kindToWatch, Runnable doExecute) {
+    private Watch watchResource(String kindToWatch, Runnable doExecute) throws IOException {
         if (kindToWatch.equalsIgnoreCase(KIND_FUNCTION)) {
             return new FunctionWatcher(kn).doWatch(doExecute);
         }
         return null;
     }
 
     public void watchResource(String id, String kindToWatch, BiConsumer<Watcher.Action, Service> doExecute) {
-        if (!watches.containsKey(id)) {
-            Watch watch = watchResource(kindToWatch, doExecute);
-            if (watch != null) {
-                watches.put(id, watch);
+        if (!watches.containsKey(id)
+                && (!watchRetry.containsKey(id) || watchRetry.get(id) < 3)) {
```

Comment:
Would me be to refactor in a separate method (eg isRetryable) to improve readbility